File: typescript/src/infrastructure/in-memory-todo-repository.ts
import type { TodoRepository } from '../domain/todo-repository.js'
import type { Todo } from '../domain/todo.js'
import type { TodoId } from '../domain/todo-id.js'

export class InMemoryTodoRepository implements TodoRepository {
  private readonly todos = new Map<string, Todo>()
  private readonly userTodos = new Map<string, Set<string>>()

  async save(todo: Todo): Promise<void> {
    const id = todo.toJSON().id
    this.todos.set(id, todo)

    const userId = todo.userId
    let userSet = this.userTodos.get(userId)
    if (!userSet) {
      userSet = new Set<string>()
      this.userTodos.set(userId, userSet)
    }
    userSet.add(id)
  }

  async findById(id: TodoId, userId: string): Promise<Todo | null> {
    const todo = this.todos.get(id.toString())
    if (todo && todo.userId === userId) {
      return todo
    }
    return null
  }

  async findAll(userId: string): Promise<Todo[]> {
    const userSet = this.userTodos.get(userId)
    if (!userSet) {
      return []
    }

    const result: Todo[] = []
    for (const id of userSet) {
      const todo = this.todos.get(id)
      if (todo && todo.userId === userId) {
        result.push(todo)
      }
    }
    return result
  }

  async delete(id: TodoId, userId: string): Promise<void> {
    const todo = this.todos.get(id.toString())
    if (todo && todo.userId === userId) {
      this.todos.delete(id.toString())

      const userSet = this.userTodos.get(userId)
      if (userSet) {
        userSet.delete(id.toString())
        if (userSet.size === 0) {
          this.userTodos.delete(userId)
        }
      }
    }
  }
}

File: csharp/src/TodoApi/Repositories/InMemoryTodoRepository.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using TodoApi.Models;

namespace TodoApi.Repositories;

public class InMemoryTodoRepository : ITodoRepository
{
    private readonly ConcurrentDictionary<string, Todo> _todos = new();
    private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, byte>> _userTodos = new();
    private readonly ConcurrentDictionary<string, string> _knownUserIdForTodo = new();

    public void AddTodo(Todo todo)
    {
        _todos[todo.Id] = todo;
        _knownUserIdForTodo[todo.Id] = todo.UserId;

        var userSet = _userTodos.GetOrAdd(todo.UserId, _ => new ConcurrentDictionary<string, byte>());
        userSet[todo.Id] = 1;
    }

    public void DeleteTodo(string id)
    {
        if (_todos.TryRemove(id, out _))
        {
            if (_knownUserIdForTodo.TryRemove(id, out var userId))
            {
                if (_userTodos.TryGetValue(userId, out var userSet))
                {
                    userSet.TryRemove(id, out _);
                }
            }
        }
    }

    public Todo? GetTodoById(string id)
    {
        _todos.TryGetValue(id, out var todo);
        return todo;
    }

    public List<Todo> GetTodosByUserId(string userId)
    {
        if (_userTodos.TryGetValue(userId, out var userSet))
        {
            var result = new List<Todo>(userSet.Count);
            foreach (var todoId in userSet.Keys)
            {
                if (_todos.TryGetValue(todoId, out var todo))
                {
                    result.Add(todo);
                }
            }
            return result;
        }

        return new List<Todo>();
    }

    public void UpdateTodo(Todo todo)
    {
        _todos[todo.Id] = todo;

        if (_knownUserIdForTodo.TryGetValue(todo.Id, out var oldUserId))
        {
            if (oldUserId != todo.UserId)
            {
                // Remove from old user's set
                if (_userTodos.TryGetValue(oldUserId, out var oldUserSet))
                {
                    oldUserSet.TryRemove(todo.Id, out _);
                }
            }
        }

        // Add/Update to current user's set
        _knownUserIdForTodo[todo.Id] = todo.UserId;
        var userSet = _userTodos.GetOrAdd(todo.UserId, _ => new ConcurrentDictionary<string, byte>());
        userSet[todo.Id] = 1;
    }
}
